internal/server/iito.go

```diff
@@ -48,56 +48,103 @@ func (s *SOMASServer) runGiftSession() error {
+		totalRequests[id], _ = s.sanitiseTeamGiftRequests(s.clientMap[id].GetGiftRequests(), id)
+
+		if len(totalRequests[id]) == 0 {
+			delete(totalRequests, id)
+		}
+	}
+	return totalRequests
+}
+
+// TODO: Return an error?
+func (s *SOMASServer) sanitiseTeamGiftOffers(offers shared.GiftOfferDict, thisTeam shared.ClientID) (shared.GiftOfferDict, error) {
+	for team, offer := range offers {
+		if s.gameState.ClientInfos[team].LifeStatus == shared.Dead || team == thisTeam || offer == 0 {
+			delete(offers, team)
+		}
 	}
-	return giftRequestDict
+	// TODO: Cap the last offer(s) so that the sum of offers doesn't exceed total resources
+	// TODO: Maybe return some kind of helpful message if any of the above cases break?
+	return offers, nil
 }
-func (s *SOMASServer) getGiftOffers(giftRequestDict shared.GiftDict) (map[shared.ClientID]shared.GiftDict, error) {
-	giftOfferDict := map[shared.ClientID]shared.GiftDict{}
-	var err error
-	for id, client := range s.clientMap {
-		giftOfferDict[id], err = client.OfferGifts(giftRequestDict)
+
+// getGiftOffers collects all responses from clients to their requests in a map
+func (s *SOMASServer) getGiftOffers(totalRequests map[shared.ClientID]shared.GiftRequestDict) (map[shared.ClientID]shared.GiftOfferDict, error) {
+	totalOffers := map[shared.ClientID]shared.GiftOfferDict{}
+	for _, thisTeam := range getNonDeadClientIDs(s.gameState.ClientInfos) {
+		// Gather all the requests made to this team
+		requestsToThisTeam := shared.GiftRequestDict{}
+		for fromTeam, indivRequests := range totalRequests {
+			requestsToThisTeam[fromTeam] = indivRequests[thisTeam]
+		}
+
+		offer, err := s.sanitiseTeamGiftOffers(s.clientMap[thisTeam].GetGiftOffers(requestsToThisTeam), thisTeam)
 		if err != nil {
-			return giftOfferDict, err
+			// totalResponses in this case may have a bogus or meaningless value
+			return totalOffers, err
 		}
+		totalOffers[thisTeam] = offer
 	}
-	return giftOfferDict, nil
+	return totalOffers, nil
 }
-func (s *SOMASServer) getGiftAcceptance(giftOffersDict map[shared.ClientID]shared.GiftDict) (map[shared.ClientID]shared.GiftInfoDict, error) {
-	acceptedGifts := map[shared.ClientID]shared.GiftInfoDict{}
-	var err error
-
-	receivedByClientDict := make(map[shared.ClientID]shared.GiftDict)
-
-	//puts all the gifts received by a certain client accesible by the id of that client
-	for idsend := range giftOffersDict {
-		for idto := range giftOffersDict {
-			if idsend == idto {
-				continue
-			}
-			if receivedByClientDict[idsend] == nil {
-				receivedByClientDict[idsend] = make(shared.GiftDict)
-			}
-			receivedByClientDict[idsend][idto] = giftOffersDict[idto][idsend]
+
+func (s *SOMASServer) sanitiseTeamGiftResponses(responses shared.GiftResponseDict, offers shared.GiftOfferDict) (shared.GiftResponseDict, error) {
+	// Cap each response so that the an island can't accept more than it was offered.
+	for team, response := range responses {
```

Comment:
this function should also remove responses that don't have a corresponding offer. Potentially a client can simply add a gift response to their returned map.

A simple fix would be to create a clean map in this function and assign to it on line 113 along with the reason. Then return this new map